download/oci_download.go

```diff
@@ -294,44 +292,50 @@ func (d *OCIDownloader) download(ctx context.Context, m metrics.Metrics) (*downl
 }
 
 func (d *OCIDownloader) pull(ctx context.Context, ref string) (*ocispec.Descriptor, error) {
-	authHeader := make(http.Header)
-	client, err := d.getHTTPClient(&authHeader)
+	lookup := d.client.AuthPluginLookup()
+
+	plugin, err := d.client.Config().AuthPlugin(lookup)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("look up auth plugin: %w", err)
 	}
-	urlInfo, err := url.Parse(d.client.Config().URL)
+
+	d.logger.Debug("OCIDownloader: using auth plugin: %T", plugin)
+
+	resolver, err := dockerResolver(plugin, d.client.Config(), d.logger)
 	if err != nil {
 		return nil, fmt.Errorf("invalid host url %s: %w", d.client.Config().URL, err)
 	}
 
-	target := newRemoteManager(d.getResolverHost(client, urlInfo), authHeader, ref)
+	target := remoteManager{
+		resolver: resolver,
+		srcRef:   ref,
+	}
 
-	manifestDescriptor, err := oraslib.Copy(ctx, target, ref, d.store, "", oraslib.DefaultCopyOptions)
+	manifestDescriptor, err := oraslib.Copy(ctx, &target, ref, d.store, "", oraslib.DefaultCopyOptions)
 	if err != nil {
 		return nil, fmt.Errorf("download for '%s' failed: %w", ref, err)
 	}
 
 	return &manifestDescriptor, nil
 }
 
-func (d *OCIDownloader) getResolverHost(client *http.Client, urlInfo *url.URL) docker.RegistryHosts {
-	creds := d.client.Config().Credentials
-	var auth docker.Authorizer
-	if creds.Bearer == nil {
-		auth = docker.NewDockerAuthorizer(
-			docker.WithAuthClient(client),
-		)
-	} else {
-		auth = docker.NewDockerAuthorizer(
-			docker.WithAuthClient(client),
-			docker.WithAuthCreds(func(string) (string, string, error) {
-				creds := d.client.Config().Credentials
-				if creds.Bearer == nil {
-					return " ", " ", nil
-				}
+func dockerResolver(plugin rest.HTTPAuthPlugin, config *rest.Config, logger logging.Logger) (remotes.Resolver, error) {
```

Comment:
The only changes to the `rest` package API are that `Client.AuthPluginLookup()` and `Config.AuthPlugin()` are now exposed.
We basically only need to be able to access different implementations of `rest.HTTPAuthPlugin` from within the `download` package. Since all implementations are part of the `rest` package, I don't see a way to access them without exposing some kind of accessor.
Only the `rest.Config` holds references to the plugin implementations and is able to return them. Unfortunately, it's an exposed type and we can't move it to another package.